File: php-checks/src/main/java/org/sonar/php/checks/regex/EmptyStringRepetitionCheck.java
package org.sonar.php.checks.regex;

import java.util.Collections;
import org.sonar.check.Rule;
import org.sonar.plugins.php.api.tree.expression.FunctionCallTree;
import org.sonarsource.analyzer.commons.regex.RegexParseResult;
import org.sonarsource.analyzer.commons.regex.ast.DisjunctionTree;
import org.sonarsource.analyzer.commons.regex.ast.GroupTree;
import org.sonarsource.analyzer.commons.regex.ast.RegexBaseVisitor;
import org.sonarsource.analyzer.commons.regex.ast.RegexTree;
import org.sonarsource.analyzer.commons.regex.ast.RepetitionTree;
import org.sonarsource.analyzer.commons.regex.ast.SequenceTree;

@Rule(key = "S5842")
public class EmptyStringRepetitionCheck extends AbstractRegexCheck {

  private static final String MESSAGE = "Rework this part of the regex to not match the empty string.";

  @Override
  public void checkRegex(RegexParseResult regexParseResult, FunctionCallTree regexFunctionCall) {
    new Visitor().visit(regexParseResult);
  }

  private class Visitor extends RegexBaseVisitor {

    @Override
    public void visitRepetition(RepetitionTree tree) {
      RegexTree element = tree.getElement();
      if (matchEmptyString(element)) {
        reportIssue(element, MESSAGE, Collections.emptyList());
      }
    }

    private boolean matchEmptyString(RegexTree element) {
      switch (element.kind()) {
        case SEQUENCE:
          return ((SequenceTree) element).getItems().stream().allMatch(this::matchEmptyString);
        case DISJUNCTION:
          return ((DisjunctionTree) element).getAlternatives().stream().anyMatch(this::matchEmptyString);
        case REPETITION:
          return ((RepetitionTree) element).getQuantifier().getMinimumRepetitions() == 0;
        case LOOK_AROUND:
        case BOUNDARY:
          return true;
        default:
          if (element instanceof GroupTree) {
            RegexTree nestedElement = ((GroupTree) element).getElement();
            return nestedElement == null || matchEmptyString(nestedElement);
          }
          return false;
      }
    }

  }
}

File: php-checks/src/test/resources/checks/regex/EmptyStringRepetitionCheck.php
<?php

namespace checks\regex;

class EmptyStringRepetitionCheck
{

  function noncompliant($input)
  {
    preg_match("/(?:)*/", $input); // Noncompliant {{Rework this part of the regex to not match the empty string.}}
    //           ^^^^
    preg_match("/(?:)?/", $input); // Noncompliant
    preg_match("/(?:)+/", $input); // Noncompliant
    preg_match("/()*/", $input); // Noncompliant
    preg_match("/()?/", $input); // Noncompliant
    preg_match("/()+/", $input); // Noncompliant
    preg_match("/xyz|(?:)*/", $input); // Noncompliant
    preg_match("/(?:|x)*/", $input); // Noncompliant
    preg_match("/(?:x|)*/", $input); // Noncompliant
    preg_match("/(?:x|y*)*/", $input); // Noncompliant
    preg_match("/(?:x*|y*)*/", $input); // Noncompliant
    preg_match("/(?:x?|y*)*/", $input); // Noncompliant
    preg_match("/(?:x*)*/", $input); // Noncompliant
    preg_match("/(?:x?)*/", $input); // Noncompliant
    preg_match("/(?:x*)?/", $input); // Noncompliant
    preg_match("/(?:x?)?/", $input); // Noncompliant
    preg_match("/(?:x*)+/", $input); // Noncompliant
    preg_match("/(?:x?)+/", $input); // Noncompliant
    preg_match("/(x*)*/", $input); // Noncompliant
    preg_match("/((x*))*/", $input); // Noncompliant
    preg_match("/(?:x*y*)*/", $input); // Noncompliant
    preg_match("/(?:())*/", $input); // Noncompliant
    preg_match("/(?:(?:))*/", $input); // Noncompliant
    preg_match("/((?i))*/", $input); // Noncompliant
    preg_match("/(())*/", $input); // Noncompliant
    preg_match("/(()x*)*/", $input); // Noncompliant
    preg_match("/(()|x)*/", $input); // Noncompliant
    preg_match("/($)*/", $input); // Noncompliant
    preg_match("/(\b)*/", $input); // Noncompliant
    preg_match("/((?!x))*/", $input); // Noncompliant
  }

  function compliant($input)
  {
    preg_match("/x*|/", $input);
    preg_match("/x*|/", $input);
    preg_match("/x*/", $input);
    preg_match("/x?/", $input);
    preg_match("/(?:x|y)*/", $input);
    preg_match("/(?:x+)+/", $input);
    preg_match("/(?:x+)*/", $input);
    preg_match("/(?:x+)?/", $input);
    preg_match("/((x+))*/", $input);
  }

  function no_duplications($input)
  {
    $regex = "/(?:)*/"; // Noncompliant
    //         ^^^^
    preg_match($regex, $input);
    preg_match($regex, $input);
  }

}
